staging/src/k8s.io/component-base/featuregate/feature_gate.go

```diff
@@ -564,15 +564,19 @@ func (f *featureGate) featureSpec(key Feature) (FeatureSpec, error) {
 }
 
 func (f *featureGate) unsafeRecordQueried(key Feature) {
-	queriedFeatures := map[Feature]struct{}{}
-	for k := range f.queriedFeatures.Load().(map[Feature]struct{}) {
-		queriedFeatures[k] = struct{}{}
-	}
-	if _, ok := queriedFeatures[key]; ok {
+	queriedFeatures := f.loadQueriedFeatures()
+	if queriedFeatures.Has(key) {
 		return
 	}
-	queriedFeatures[key] = struct{}{}
-	f.queriedFeatures.Store(queriedFeatures)
+	queriedFeatures.Insert(key)
```

Comment:
I'd expect to see the set cloned here before any mutation happens, does something else make this safe?

Reply:
`Enabled` is not a thread safe function. Using the pointer could result in race conditions to update the set. But since `queriedFeatures` is only used to print out warning message, I guess this is ok?

Reply:
I would expect the runtime to panic with `concurrent map read and write` here without a copy or some kind of mutex.

Reply:
Please see
```
func (f *featureGate) Enabled(key Feature) bool {
	// TODO: ideally we should lock the feature gate in this call to be safe, need to evaluate how much performance impact locking would have.
```

Reply:
Good idea, updated to make it clone the set only when mutation is necessary.

`Enabled` already has a TODO to make it thread-safe, so I also added locking/unlocking

Reply:
> clone the set only when mutation is necessary

This is enough to resolve my concern. Cloning was still necessary in the mutation case to avoid panics. Eliminating the clone for the non-mutation case seems to be the essential change that would reduce allocs.